hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieDataSourceHelper.scala

```diff
@@ -0,0 +1,197 @@
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat
+import org.apache.spark.sql.execution.datasources.{FilePartition, PartitionedFile}
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+
+object HoodieDataSourceHelper extends PredicateHelper {
+
+  /**
+   * Partition the given condition into two sequence of conjunctive predicates:
+   * - predicates that can be evaluated using metadata only.
+   * - other predicates.
+   */
+  def splitPartitionAndDataPredicates(
+      spark: SparkSession,
+      condition: Expression,
+      partitionColumns: Seq[String]): (Seq[Expression], Seq[Expression]) = {
+    splitConjunctivePredicates(condition).partition(
+      isPredicateMetadataOnly(spark, _, partitionColumns))
+  }
+
+  /**
+   * Check if condition can be evaluated using only metadata. In Delta, this means the condition
+   * only references partition columns and involves no subquery.
+   */
+  def isPredicateMetadataOnly(
+      spark: SparkSession,
+      condition: Expression,
+      partitionColumns: Seq[String]): Boolean = {
+    isPredicatePartitionColumnsOnly(spark, condition, partitionColumns) &&
+        !SubqueryExpression.hasSubquery(condition)
+  }
+
+  /**
+   * Does the predicate only contains partition columns?
+   */
+  def isPredicatePartitionColumnsOnly(
+      spark: SparkSession,
+      condition: Expression,
+      partitionColumns: Seq[String]): Boolean = {
+    val nameEquality = spark.sessionState.analyzer.resolver
+    condition.references.forall { r =>
+      partitionColumns.exists(nameEquality(r.name, _))
+    }
+  }
+
+  /**
+   * Wrapper `buildReaderWithPartitionValues` of [[ParquetFileFormat]]
+   * to deal with [[ColumnarBatch]] when enable parquet vectorized reader if necessary.
+   */
+  def buildHoodieParquetReader(
+      sparkSession: SparkSession,
+      dataSchema: StructType,
+      partitionSchema: StructType,
+      requiredSchema: StructType,
+      filters: Seq[Filter],
+      options: Map[String, String],
+      hadoopConf: Configuration): (PartitionedFile) => Iterator[InternalRow] = {
+
+    val readParquetFile: PartitionedFile => Iterator[Any] = new ParquetFileFormat().buildReaderWithPartitionValues(
+      sparkSession = sparkSession,
+      dataSchema = dataSchema,
+      partitionSchema = partitionSchema,
+      requiredSchema = requiredSchema,
+      filters = filters,
+      options = options,
+      hadoopConf = hadoopConf
+    )
+
+    file: PartitionedFile => {
+      val iter = readParquetFile(file)
+      unravelColumnarBatchIfNecessary(iter)
+    }
+  }
+
+  /**
+   * if is [[ColumnarBatch]], unravel it to [[InternalRow]]
+   */
+  def unravelColumnarBatchIfNecessary(iter: Iterator[Any]): Iterator[InternalRow] = {
```

Comment:
I don't think this should be used at all outside of building the reader

Reply:
let move it into `buildHoodieParquetReader`